filestore/filepath.go

```diff
@@ -0,0 +1,544 @@
+	// Scheme encompasses
+	// * protocol (e.g. s3://, gs://, abfss://)
+	// * host (e.g. <container>@<account>.dfs.core.windows.net for Azure Blob)
+	// * port (if applicable)
+	// This naming technically conflicts with the standard definition of "scheme," which
+	// is only the protocol _without_ the domain and port (i.e. authority); however, given
+	// we're not using these components independently and there's no single term to denote
+	// <SCHEME>://<HOST>:<PORT>, we're using "scheme" to encompass all three.
+	Scheme() string
+	SetScheme(scheme string) error
+
+	// Returns the name of the bucket (S3) or container (Azure Blob Storage)
+	Bucket() string
+	SetBucket(bucket string) error
+
+	// Returns the blob key, which is the relative path to the object (i.e. without the scheme or bucket/container)
+	Key() string
+	SetKey(key string) error
+
+	// Returns the key prefix (i.e. the directory path to the object)
+	KeyPrefix() string
+
+	IsDir() bool
+	SetIsDir(isDir bool)
+
+	// Returns the file extension (e.g. "parquet", "csv", etc. of the object)
+	Ext() FileType
+
+	// Returns the full path to the object, including the scheme and bucket/container
+	ToURI() string
+	// Consumes a URI (e.g. abfss://<container>@<storage_account>/path/to/file) and parses it into
+	// the specific parts that the implementation expects.
+	ParseFilePath(path string) error
+	ParseDirPath(path string) error
+
+	Validate() error
+	IsValid() bool
+}
+
+func NewEmptyFilepath(storeType FileStoreType) (Filepath, error) {
+	switch storeType {
+	case S3:
+		return &S3Filepath{FilePath{isDir: false}}, nil
+	case Azure:
+		return &AzureFilepath{}, nil
+	case GCS:
+		return &GCSFilepath{FilePath{isDir: false}}, nil
+	case Memory:
+		return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	case FileSystem:
+		return &LocalFilepath{FilePath{isDir: false}}, nil
+	//case DB:
+	//	return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	case HDFS:
+		return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	default:
+		return nil, fmt.Errorf("unknown store type '%s'", storeType)
+	}
+}
+
+func NewEmptyDirpath(storeType FileStoreType) (Filepath, error) {
+	switch storeType {
+	case S3:
+		return &S3Filepath{FilePath{isDir: true}}, nil
+	case Azure:
+		return &AzureFilepath{}, nil
+	case GCS:
+		return &GCSFilepath{FilePath{isDir: true}}, nil
+	case Memory:
+		return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	case FileSystem:
+		return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	//case DB:
+	//	return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	case HDFS:
+		return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	default:
+		return nil, fmt.Errorf("unknown store type '%s'", storeType)
+	}
+}
+
+type FilePath struct {
+	scheme  string
+	bucket  string
+	key     string
+	isDir   bool
+	isValid bool
```

Comment:
not convinced we need this

Reply:
something feels off and this design and I don't have a good solution yet. All I know is that it's a bit confusing.

correct me if i'm wrong but the way it will always work is
- you always create a filepath like this -> filestore.NewEmptyFilepath(store.FilestoreType()).
- and you do this so that you can eventually parse some path with the fileStore's schema

so I'm thinking if this is the use case, then the way things should be done is, store.ParseFilePathFromUrl() or store.ParseFilePathFromKey().